tests/Common/ManagerTestLogicTrait.php

```diff
@@ -0,0 +1,878 @@
+        $manager->canAddChild('author', 'reader');
+    }
+
+    public function testAddChild(): void
+    {
+        $manager = $this->createFilledManager();
+        $returnedManager = $manager->addChild('reader', 'createPost');
+
+        $this->assertEqualsCanonicalizing(
+            [
+                'readPost',
+                'createPost',
+            ],
+            array_keys($this->itemsStorage->getChildren('reader'))
+        );
+        $this->assertSame($manager, $returnedManager);
+    }
+
+    public function testAddChildNotHasItem(): void
+    {
+        $manager = $this->createFilledManager();
+
+        $this->expectException(InvalidArgumentException::class);
+        $this->expectExceptionMessage('Either "new reader" does not exist.');
+
+        $manager->addChild(
+            'new reader',
+            'createPost'
+        );
+    }
+
+    public function testAddChildEqualName(): void
+    {
+        $manager = $this->createFilledManager();
+
+        $this->expectException(InvalidArgumentException::class);
+        $this->expectExceptionMessage('Cannot add "createPost" as a child of itself.');
+
+        $manager->addChild(
+            'createPost',
+            'createPost'
+        );
+    }
+
+    public function testAddChildPermissionToRole(): void
+    {
+        $manager = $this->createFilledManager();
+
+        $this->expectException(InvalidArgumentException::class);
+        $this->expectExceptionMessage('Can not add "reader" role as a child of "createPost" permission.');
+
+        $manager->addChild(
+            'createPost',
+            'reader'
+        );
+    }
+
+    public function testAddChildAlreadyAdded(): void
+    {
+        $manager = $this->createFilledManager();
+
+        $this->expectException(RuntimeException::class);
+        $this->expectExceptionMessage('The item "reader" already has a child "readPost".');
+
+        $manager->addChild(
+            'reader',
+            'readPost'
+        );
+    }
+
+    public function testAddChildDetectLoop(): void
+    {
+        $manager = $this->createFilledManager();
+
+        $this->expectException(RuntimeException::class);
+        $this->expectExceptionMessage('Cannot add "author" as a child of "reader". A loop has been detected.');
+
+        $manager->addChild(
+            'reader',
+            'author',
+        );
+    }
+
+    public function testAddChildWithNonExistChild(): void
+    {
+        $manager = $this->createFilledManager();
+
+        $this->expectException(InvalidArgumentException::class);
+        $this->expectExceptionMessage('Either "new reader" does not exist.');
+        $manager->addChild('reader', 'new reader');
+    }
+
+    public function testRemoveChild(): void
+    {
+        $manager = $this->createFilledManager();
+        $returnedManager = $manager->removeChild('author', 'createPost');
+
+        $this->assertEqualsCanonicalizing(
+            [
+                'readPost',
```

Comment:
```suggestion
```

`readPost` is nested children, `getChildren` returns directly childrens only.

Reply:
Current DB implementations use recursion both for getting parents and children:

https://github.com/yiisoft/rbac-db/blob/75a80e62702b32d593602dc20844b2963de400b6/src/ItemsStorage.php#L225-L235

https://github.com/yiisoft/rbac-cycle-db/blob/d94adbf7eb7d3c47cabb032c4e366f1b481e67e1/src/ItemsStorage.php#L242-L252

Then should it be fixed there too?

Reply:
https://github.com/yiisoft/rbac/pull/161#discussion_r1278511215

Reply:
https://github.com/yiisoft/rbac/pull/161#discussion_r1285415201